lib/syskit/test/execution_expectations.rb

```diff
@@ -21,26 +21,48 @@ def self.resolve_orocos_reader(reader)
                 reader
             end
 
+            class HaveNoNewSample < Roby::Test::ExecutionExpectations::Maintain
+                def initialize(reader, at_least_during, backtrace)
+                    @reader = reader
+                    orocos_reader = ExecutionExpectations.resolve_orocos_reader(reader)
+                    block = proc { !(@received_sample = orocos_reader.read_new) }
+                    super(at_least_during, block, "", backtrace)
+                end
+
+                def to_s
+                    "#{@reader} should not have received a new sample"
+                end
+
+                def explain_unachievable(propagation_info)
+                    @received_sample
+                end
+
+                def format_unachievable_explanation(pp, explanation)
+                    pp.text "but it received one: "
+                    explanation.pretty_print(pp)
+                end
+            end
+
             # Expect that no new samples arrive on the reader for a certain time
             # period
             #
             # @param [Float] at_least_during no samples should arrive for at
             #   least that many seconds. This is a minimum.
             # @return [nil]
-            def have_no_new_sample(reader, at_least_during: 0)
-                reader = ExecutionExpectations.resolve_orocos_reader(reader)
-                maintain(at_least_during: at_least_during, description: "#{reader} has one new sample #{sample}, but none was expected") do
-                    !reader.read_new
-                end
+            def have_no_new_sample(reader, at_least_during: 0, backtrace: caller(1))
+                add_expectation(HaveNoNewSample.new(
+                    reader, at_least_during, backtrace))
                 nil
             end
 
             # Expect that one sample arrives on the reader, and return the sample
             #
             # @return [Object]
             def have_one_new_sample(reader)
-                reader = ExecutionExpectations.resolve_orocos_reader(reader)
-                achieve(description: "receive one sample on #{reader}") { reader.read_new }
+                orocos_reader = ExecutionExpectations.resolve_orocos_reader(reader)
```

Comment:
We should give the backtrace as argument to all the expectation, otherwise the failure backtrace is going to be the one of the nested expectation and not from the caller.

Reply:
Done